platform-sdk/docs/proposals/healthMonitor/healthMonitorProposal.md

```diff
@@ -0,0 +1,127 @@
+# Health Monitor Proposal
+
+---
+
+## Summary
+
+A replacement for the current hard backpressure mechanism using a health monitor.
+
+| Metadata           | Entities                            | 
+|--------------------|-------------------------------------|
+| Designers          | Cody Littley                        |
+| Functional Impacts | Consensus node performance          |
+| Related Proposals  | n/a                                 |
+| HIPS               | n/a                                 |
+
+---
+
+## Purpose and Context
+
+The platform's current backpressure mechanism has very bad performance characteristics.
+We need to replace the current backpressure implementation with something more performant.
+
+### Requirements
+
+- The backpressure mechanism must prevent an OOM in the event of any arbitrary component
+  becoming arbitrarily slow.
+- The backpressure mechanism should not slow down critical processes when enganged.
+  Specifically, transaction handling should not be impeded when backpressure is engaged.
+  This includes slowdowns due to the transaction handling thread being starved of work.
+- The backpressure mechanism should not be overly aggressive. That is, we shouldn't see
+  it engaging under normal circumstances on nodes that are not in a state of distress.
+- The implementation of the backpressure mechanism should be sufficiently clean and
+  maintainable. A performant solution that is too complex is not a solution we should
+  accept as a solution.
+
+### Design Decisions
+
+This change can be split into two distinct sub-problems: the detection of a system in duress, and the action that
+is taken to apply backpressure on a distressed system.
+
+#### Detection
+
+The current platform is built from a number of "components". A component is usually a unit of logic that runs on its
+own thread (or logical thread). Each component has a queue (or logical queue) of data it is ingesting. If a component
+produces output, that output is enqueued within the queues of other components.
+
+For each component with a queue, define a preferred size. The queue will not apply hard backpressure even if the
+size of the queue exceeds the preferred size. That is, insertion into the queue will never block. However, a queue
+whose size exceeds its preferred size will be considered to be "unhealthy".
```

Comment:
The fundamental problem I have this proposal is this line: "define a preferred size". It may be that any solution we devise has some metric that must be defined. The fewer of these we have, the better. We should strive for a design with the fewest knobs possible. The more knobs, the more likely we get one wrong, leading to aberrant behavior.

I think we should be thinking about larger "chunks". The current breakdown of components and health checking between components is too fine-grained. Jasper also opened my mind to another problem we really need to think about: each task added to the fork-join pool may take an unknown amount of time to execute. Or a single component might put lots and lots of thing son the FJP. In both cases, we have the potential for a FJP DOS. Even if we went to the trouble of adopting this design, we are likely to find it doesn't actually help, because the FJP is mismatched on # worker threads and type and shape of tasks added to the FJP.

Reply:
> we are likely to find it doesn't actually help

Initial testing shows significant improvement in the health of the system.